### PR TITLE
`General`: Redesign the course overview

### PR DESCRIPTION
**Pull request by edkaya**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Since the current design of the course overview page lacks user-friendliness, the pre-designed course cards will be im-
plemented to make the page appear cleaner and more structured. This implementation is a part of the new design for course overview page.

### Description

- Changed course card sizes and border design 
- Changed the way of displaying information of the score and next exercise
- Added Manage Course button for tutors, editors and instructors to navigate to the corresponding course management page

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Tutor || Admin || Editor || Instructor

1. Log in to Artemis with student account
2. Verify that the course cards look good 
3. Verify that the course cards are responsive in case of size changes of page in width
4. Verify that you can navigate to the exercise by clicking on the exercise title under the "Next Exercise"
5. Log in to Artemis with Tutor || Admin || Editor || Instructor account
6. Verify that you can see manage course button 
7. Verify that manage course button works as expected

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Before 
Screenshot 2024-08-15 at 20 18 48

After 
Screenshot 2024-08-23 at 17 51 34

## Summary by CodeRabbit

- **New Features**
	- Enhanced course management interface with a new management button for direct access.
	- Introduced new methods to manage the state of the course overview.

- **User Interface Improvements**
	- Modernized the layout and design of the course card component with improved flexibility and visual clarity.
	- Enhanced pie chart representation and refined statistics display logic for better user feedback.
	- Improved responsive design across various screen sizes for better usability.
	- Updated HTML structure in the course overview component for better visual distinction and modernity.
	- Introduced a grid layout for course cards, enhancing adaptability and aesthetics.
	- Enhanced localization for both German and English versions of the student dashboard.

- **Bug Fixes**
	- Resolved issues related to state management of the course overview display.

### CODE REVIEW

File: src/main/webapp/app/overview/course-card.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/course-card.component.html:29]: Please add a translation for this

File: src/main/webapp/i18n/de/student-dashboard.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/student-dashboard.json:15]:
```suggestion
            "cardNextUnit": "Nächste Aufgabe",
```

File: src/main/webapp/i18n/en/student-dashboard.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/student-dashboard.json:15]:
```suggestion
            "cardNextUnit": "Next Exercise",
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Comment by b-fein** [comment]
Design comments looking at the screenshots:

- The headers of the cards now look like they take up more space than before. Would it be possible to make them a bit smaller again to reduce amount of blank space on the page?
- Artemis uses fairly rectangular shapes consistently throughout the UI. I don’t think the very rounded corners fit into this design. In the past there were even dedicated redesigns to make everything match the rectangular aesthetic (especially lectures and tutorials were not matching the other parts of the app if I remember correctly).

File: src/main/webapp/app/overview/course-card.component.html
Comment (pzdr7) [inline src/main/webapp/app/overview/course-card.component.html:81]: Aren't instructors and editors already "at least tutors"?

```suggestion
            @if (course.isAtLeastTutor) {
```

File: src/main/webapp/app/shared/layouts/main/main.component.scss
Comment (pzdr7) [inline src/main/webapp/app/shared/layouts/main/main.component.scss:3]:
```suggestion
    padding-top: 0;
```

or add the bootstrap class `pt-0` to the affected element

File: src/main/webapp/app/overview/course-card.component.html
Comment (pzdr7) [inline src/main/webapp/app/overview/course-card.component.html:57]: You might be able to simplify this by leaving out `> 0`, e.g. for exerciseCount. Can you check?

**Review by pzdr7 — CHANGES_REQUESTED** [review]
Some small comments on the code

----

Regarding the design:

image

It seems that the image offsets the position of the course name. I'd prefer it if this wasn't the case. Maybe we can align the text differently and/or add a placeholder for courses without an image?

**Comment by edkaya** [comment]
> Design comments looking at the screenshots:
> 
>     * The headers of the cards now look like they take up more space than before. Would it be possible to make them a bit smaller again to reduce amount of blank space on the page?
> 
>     * Having the vertical and horizontal spacing in the grid being different looks a bit weird to me. I think I’d prefer a consistent spacing like between rows at the moment also be used between items within the row.
> 
>     * Artemis uses fairly rectangular shapes consistently throughout the UI. I don’t think the very rounded corners fit into this design. In the past there were even dedicated redesigns to make everything match the rectangular aesthetic (especially lectures and tutorials were not matching the other parts of the app if I remember correctly).

Thanks a lot for your insights! We have adjusted design according to your remarks, which make sense in general. That's what we have changed: 
- The height of the header is reduced 
- We have more consistent spacing between the cards. Especially between the the header, cards, and the borders. Since we don't want to expand the cards too much like the current version (it happens if you have a larger screen and the grid tries to fit the content), we have a little bit more spacing between cards, which is a trade-off but expanding the card looks off and not good in general. 
- We reduced the rounded corners and adjusted to make it look consistent throughout Artemis.

**Comment by janthoXO** [comment]
might be just a little detail, but isnt the green circle slightly slimmer than the red one
image

**Review by janthoXO — CHANGES_REQUESTED** [review]
1. The "Next Exercise" does not communicate that it is clickable. As it has the same font and style as the points text and any "normal" text on the website

2. If the solution for 1. is to use the same style as "Manage Course" then the course card might look overloaded with text buttons. therefore I would change the "manage course" button to an icon button. probably something like an edit button with an pen icon.
In addition you should probably move this icon then somewhere to the top right of the card

image


Student View:
the padding on the bottom looks a bit too much. I know that's needed for the "manage course" button in instructor view but still looks a bit off. Using the icon pen button as said before, would solve this issue as well, as the padding could be reduced in both views
image

**Review by anian03 — CHANGES_REQUESTED** [review]
In general I do quite like the changes, great job! Just two things I noticed that we might want to change:

1. There appears to be some sort of double border (with non-matching corner radii) in dark mode, which looks strange
Double border

2. The padding on the left and right side looks quite big if there is just one card (narrow screens like phone, Split View, …). Maybe can the cards just take up most of the width if just one is displayed per row?
Big horizontal empty space

**Comment by edkaya** [comment]
> 1. The "Next Exercise" does not communicate that it is clickable. As it has the same font and style as the points text and any "normal" text on the website
> 
>     2. If the solution for 1. is to use the same style as "Manage Course" then the course card might look overloaded with text buttons. therefore I would change the "manage course" button to an icon button. probably something like an edit button with an pen icon.
>        In addition you should probably move this icon then somewhere to the top right of the card
> 
> 
> image
> 
> Student View: the padding on the bottom looks a bit too much. I know that's needed for the "manage course" button in instructor view but still looks a bit off. Using the icon pen button as said before, would solve this issue as well, as the padding could be reduced in both views image

Thanks for your advices. You are right about the exercise title. I changed it to make it look clickable. Regarding your remaining points, we discussed these today with @rabeatwork and we came to conclusion that using an icon in the header will not be the best fit here, because it's not clear what will happen by clicking on this icon and there could be many possibilites from a user perspective. Users could understand that icon as course settings or editing the course title or something similar. Using a label like "Manage Course" will definetely convey the intended message here. Besides that, we may have some plans about the design of the header, and we can incorporate banner images. We'll need to consider the icon's visibility against those images, too.

**Comment by edkaya** [comment]
> In general I do quite like the changes, great job! Just two things I noticed that we might want to change:
> 
>     1. There appears to be some sort of double border (with non-matching corner radii) in dark mode, which looks strange
> 
> 
> Double border
> 
>     2. The padding on the left and right side looks quite big if there is just one card (narrow screens like phone, Split View, …). Maybe can the cards just take up most of the width if just one is displayed per row?
> 
> 
> Big horizontal empty space
> 
> Agree with the previous comment that it would be good to know that the next exercise is clickable – either via color and/or an arrow like "Manage Course".

Thanks, I fixed the border issue now. Regarding the spacing for one card, I think it will not lead to an issue on narrow screens, since this padding will not appear in these screens. You are right about the split view, but it seems an edge case to us, where the user views these cards in split view with only one card, because users generally use these split views after navigating to the course to see actual course content rather than course cards. Since we want to maintain consistency for the course card sizes across diferent screen sizes, we kind of make a trade-off here.

**Review by florian-glombik — CHANGES_REQUESTED** [review]
The code looks quite clean, I only left some minor comments and suggestions, for my approve please address the comment in `src/main/webapp/app/overview/courses.component.html`

**Review by SimonEntholzer — APPROVED** [review]
The only thing that could be maybe improved is the overall course view layout when its very zoomed out, as the distances from the course cards increase a lot:
image

But since people usually do not zoom out that far and don't even have that many courses, I think it's fine.

File: src/main/webapp/app/course/manage/course-management.service.ts
Comment (pzdr7) [inline src/main/webapp/app/course/manage/course-management.service.ts:705]: Maybe `enable` and `disable` would be a better terminology here than set and reset

File: src/main/webapp/app/overview/course-card.component.html
Comment (pzdr7) [reply src/main/webapp/app/overview/course-card.component.html:59]: I think for `course.id` this is a problem of the template in general, rather than this specific location - basically the entire template relies on `course.id` not being `undefined` (everything but the outermost div). Personally I'd be fine with leaving the assertion for `course.id` as is. No strong opinion either way

**Review by pzdr7 — COMMENTED** [review]
I can second what @SimonEntholzer said about the spacing of the cards. I have a very wide screen, which makes the cards look a bit small compared to the gaps/overall space. Probably an edge case, though

image

**Review by anian03 — COMMENTED** [review]
I think there is still something not matching with the inner and outer corner radius (see the black thing in the corner). That's not a big deal though, since it's not very noticeable. Will approve once the other comments are resolved :)

**Comment by edkaya** [comment]
> sometimes the course overview loads extremly long. I'm not sure if its related to this PR but I never noticed it before so it might be related
> Screen.Recording.2024-08-28.at.10.19.41.mov

I don't have that issue when navigating to the course overview page. Are some of the other devs also experiencing this issue?

**Comment by pzdr7** [comment]
> > sometimes the course overview loads extremly long. I'm not sure if its related to this PR but I never noticed it before so it might be related
> > Screen.Recording.2024-08-28.at.10.19.41.mov
> 
> I don't have that issue when navigating to the course overview page. Are some of the other devs also experiencing this issue?

I couldn't reproduce this issue locally (tried develop & this branch). On TS3 (with admin access), there does appear to be some load time - but mostly on the course management side.

image

**Comment by edkaya** [comment]
> I can second what @SimonEntholzer said about the spacing of the cards. I have a very wide screen, which makes the cards look a bit small compared to the gaps/overall space. Probably an edge case, though
> 
> image

Thanks for your insights regarding the layout @pzdr7 and @SimonEntholzer. I added some breakpoints to adjust card sizes for larger screen to omit these edge cases as much as possible now 👍

**Review by pzdr7 — DISMISSED** [review]
Approve code 👍

**Review by anian03 — DISMISSED** [review]
Looks good to me :D

**Review by SimonEntholzer — COMMENTED** [review]
One other thing I noticed: When going to the page, it first does not show the course pictures, and displays the title at the top.

https://github.com/user-attachments/assets/97e3bbc3-ef36-446f-a393-4e41e5ca8c3e

Would it be possible, to display the text at the correct place from the beginning?
And (or), like some websites do it, use some placeholder until the content has loaded, e.g. similar like this: https://experience.sap.com/fiori-design-web/placeholder-loading/

File: src/main/webapp/app/overview/course-card.scss
Comment (florian-glombik) [inline src/main/webapp/app/overview/course-card.scss:17]: Are these values based on trial and error? Do they have a deeper meaning?

Would it make sense to add an explanation here?

File: src/main/webapp/app/overview/courses.component.html
Comment (florian-glombik) [inline src/main/webapp/app/overview/courses.component.html:62]: I do not think that this comment adds a value, rather a comment for the future, does not need to be changed
```suggestion
```

**Review by florian-glombik — COMMENTED** [review]
The code looks good to me except for `src/main/webapp/app/overview/course-card.scss` where it is not always clear to me on which basis we are choosing the values (e.g. `height: 283px` seems like a value that is inferred from an element, from which element?)

Could you have another look at this file and double check if it would make sense to add more comments on how we get to the px values / what they are meant to mean / supposed to achieve? 

If there are no useful comments that can be added I am happy to approve :)

File: src/main/webapp/app/overview/course-card.scss
Comment (edkaya) [reply src/main/webapp/app/overview/course-card.scss:17]: Not needed anymore, will remove them 👍

**Comment by edkaya** [comment]
> One other thing I noticed: When going to the page, it first does not show the course pictures, and displays the title at the top.
> 
>  Screencast.from.2024-08-29.10-15-40.mp4 
> Would it be possible, to display the text at the correct place from the beginning? And (or), like some websites do it, use some placeholder until the content has loaded, e.g. similar like this: https://experience.sap.com/fiori-design-web/placeholder-loading/

I think this happens due to high number of courses in test servers. I think it won't be a real case, since a student can have ideally 8-9 courses at most.

**Review by anian03 — APPROVED** [review]
Tested again on TS3, works as expected

**Review by pzdr7 — DISMISSED** [review]
Reapprove after reviews

**Review by florian-glombik — DISMISSED** [review]
Approve code, thanks for the changes!

**Comment by janthoXO** [comment]
> I think there is still something not matching with the inner and outer corner radius (see the black thing in the corner). That's not a big deal though, since it's not very noticeable. Will approve once the other comments are resolved :) 

I would also say that this and the spacing on wide screens are the only issues left. After that you'll get an approve

**Comment by edkaya** [comment]
> > I think there is still something not matching with the inner and outer corner radius (see the black thing in the corner). That's not a big deal though, since it's not very noticeable. Will approve once the other comments are resolved :) 
> 
> I would also say that this and the spacing on wide screens are the only issues left. After that you'll get an approve

I fixed the border issue 👍 I normally expanded the cards for wide screens but I recently got some feedback from Ramona  that we don't want to expand the cards. Instead, we now display more cards in a row if the screen is larger. But I don't think we can have a super optimal solution for very wide screens (+3500 pixels) to decrease that spacing, since it is the intended behaviour now rather than an issue.

**Review by anian03 — APPROVED** [review]
Thanks for the changes! Tested & looks good

**Review by janthoXO — CHANGES_REQUESTED** [review]
The border radius looks good now

The spacing is not satisfying yet:
image

Another problem I just noticed, is that for some images the icon is not a circle but rather an elipsis
image

**Comment by edkaya** [comment]
> The border radius looks good now
> 
> The spacing is not satisfying yet: image
> 
> Another problem I just noticed, is that for some images the icon is not a circle but rather an elipsis image

Thanks for your response. Regarding the spacing again, I now added some additional breakpoints for very large screens. We now display 9 cards at most for ~5000 pixels width, so the spacing should be decreased and fine now. I don't think we should take care of the case where the display is more than 5000 pixels, because the cards will be barely visible. 
Screenshot 2024-09-02 at 17 35 28

Regarding the icon issue, this is actually not related to this PR and I think something is wrong with the icon itself. Because it is also ellipsis in the current course card version: 
Screenshot 2024-09-02 at 17 02 59

**Review by anian03 — DISMISSED** [review]
Re-approve, still lgtm on TS3

**Review by SimonEntholzer — DISMISSED** [review]
I think it looks good now :+1: tested on TS3

**Comment by janthoXO** [comment]
> Thanks for your response. Regarding the spacing again, I now added some additional breakpoints for very large screens. We now display 9 cards at most for ~5000 pixels width, so the spacing should be decreased and fine now. I don't think we should take care of the case where the display is more than 5000 pixels, because the cards will be barely visible. 

### Spacing
I dont quite understand yet why we "hardcode" the break points. Is there no way in angular to just fit as many as possible in one row. Im only used to Flutter where there is a [Wrap Component](https://api.flutter.dev/flutter/widgets/Wrap-class.html) which does exactly that.
Another solution for the problem would be to hard code the maximum pixel width for the box containing all course card. so that with increasing screen width only the padding on the left and right side of the screen increases. Github does exactly that for example:
image


> Regarding the icon issue, this is actually not related to this PR and I think something is wrong with the icon itself. Because it is also ellipsis in the current course card version

even if its not directly related to the PR, as this PR already touches the course cards, we could fix the issue. 
if you really don't want to fix it in this PR could you please create an issue and fix it in a separate PR.

**Review by florian-glombik — DISMISSED** [review]
Approve code

I do not have a strong opinion on the suggestion https://github.com/ls1intum/Artemis/pull/9221#issuecomment-2325216957. If we can set a fixed max-width and a fixed padding I agree that this might be a better solution. If we need to adjust the padding at certain breakpoints I am not sure if it makes a difference.

**Review by edkaya — COMMENTED** [review]
I used a different approach to omit using breakpoints. Instead of using columns and modifying them according to the breakpoints, I implemented with grid layout, which made the code more intuitive now.

**Review by anian03 — CHANGES_REQUESTED** [review]
Hey, I just checked out the current version on my phone and noticed an issue with the layout: The course cards are wider than their container. Is it possible to fix this?
I think for these small screen sizes we could simply make the container take up the whole width, there is no advantage in having extra padding around it when there already is padding inside the container.

**Review by anian03 — APPROVED** [review]

**Comment by anian03** [comment]
Thanks for the fix! Layout looks good on mobile now 👍

**Review by SimonEntholzer — APPROVED** [review]
re-approve, looks good on ts3

**Review by janthoXO — APPROVED** [review]
looks really good now. thanks for implementing the changes

File: src/main/webapp/app/overview/course-card.scss
Comment (florian-glombik) [inline src/main/webapp/app/overview/course-card.scss:3]: 80px + 200px = 280px

What are the 3 px related to?

**Review by florian-glombik — COMMENTED** [review]
Thanks for the comments explaining what happens in the CSS, there is only one comment that I do not quite understand 😅

File: src/main/webapp/app/overview/course-card.scss
Comment (edkaya) [reply src/main/webapp/app/overview/course-card.scss:3]: That additional pixel is for the course manage button, I didnt write it explicitly because it was not the case for all cards 😅

Comment (edkaya) [reply src/main/webapp/app/overview/course-card.scss:3]: But I can change this comment too, as I will modify some css properties of the course cards in my follow up PR

**Review by florian-glombik — APPROVED** [review]
Approve code

**Review by pzdr7 — DISMISSED** [review]
Reapprove code

**Review by anian03 — APPROVED** [review]
Tested on TS3, Score percentage has been moved

**Review by SimonEntholzer — APPROVED** [review]
Looks good on Ts3, reapprove :+1:

**Review by florian-glombik — APPROVED** [review]
Approve code

**Review by pzdr7 — APPROVED** [review]
Code

**Review by janthoXO — APPROVED** [review]
tested on ts1

**Pull request merged**
